gcloud/datastore/query.py

```diff
@@ -104,7 +102,7 @@ def dataset_id(self):
 
         :rtype: str
         """
-        return self._dataset_id
+        return self._dataset_id or self._client.dataset_id
```

Comment:
I did it that way initially.  I don't really mind either way:  if we're going to leave it as a property (e.g., to ensure read-only), the overhead of a function call completely dominates the test.